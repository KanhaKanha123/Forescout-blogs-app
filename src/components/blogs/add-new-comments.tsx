import { FC, ReactElement, useState } from "react";
import { AddNewCommentsWrapper } from "./style";
import { Button, Card } from "../shared-comonents";
import { TextBox } from "../header/style";
import { debounce } from "../../helpers/helper";

export interface AddNewCommentsTypes {
    readonly blogId: number,
    readonly showHideAddNewCommentBoxHandler: () => void
};

/**
 * This component is responsible to add new comments
 * @params blogId
 * @returns ReactElement
 */
export const AddNewComments: FC<AddNewCommentsTypes> = ({ blogId, showHideAddNewCommentBoxHandler }): ReactElement => {

    const [newCommentText, setNewCommentText] = useState('');
    //on text change handler
    const onTextChange = async (e: any) => {

        //debounce for better performance
        const result = await debounce(500);

        const commentVal = e.target.value;

        setNewCommentText(commentVal);
    };

    const commentSaveHandler = () => {
        try {
            //if text box is empty just return back
            if (newCommentText === "") {
                alert('enter value to text box');
                return;
            }

            //here we can make an api call to save the new comment to db for blog id
            console.log('new comment: ', `blogId: ${blogId}, commentText: ${newCommentText}`);

            //clear the popup
            showHideAddNewCommentBoxHandler();
        } catch (err: any) {
            throw new Error(err)
        }

    }

    return <Card
        width="500px"
        height="200px"
        padding="15px"
        backgroundcolor="#f4f4f4">
        <TextBox required data-testid='comment_text' area-label='this is a add new comment textbox' type='text'
            placeholder='Add new comment'
            onChange={onTextChange} />
        <Button data-testid='save_comment' aria-label="Save new comment" backgroundcolor="#4792ef" onClick={commentSaveHandler}>Save</Button>
    </Card>
};