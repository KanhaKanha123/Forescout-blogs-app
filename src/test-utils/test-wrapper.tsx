import ContextProvider from '../store/context';
import { QueryClientProvider } from "react-query";
import { FC, ReactElement } from 'react';
import { client } from '../react-query/query-client';
import { BrowserRouter as Router } from 'react-router-dom';

interface TestWrapperProps {
    children: React.ReactNode
}

/**
 * This component is responsible to wrap the tests with all the important wrappers so all the time in all tests we don't have to repeat.
 * @returns ReactElement
 */
export const TestWrapper: FC<TestWrapperProps> = ({ children }): ReactElement => (
    <QueryClientProvider client={client}>
        <ContextProvider>
            <Router>
                {children}
            </Router>
        </ContextProvider>
    </QueryClientProvider>);