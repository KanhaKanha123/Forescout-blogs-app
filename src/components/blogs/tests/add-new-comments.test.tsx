import { fireEvent, render } from '@testing-library/react';
import { AddNewComments } from '../add-new-comments';
import { TestWrapper } from '../../../test-utils/test-wrapper';
import React from 'react';

const showHideAddNewCommentBoxHandler = jest.fn();

const rederComponent = () => render(<TestWrapper>
  <AddNewComments showHideAddNewCommentBoxHandler={showHideAddNewCommentBoxHandler} blogId={1} />
</TestWrapper>);

describe('AddNewComments component', () => {

  test('renders AddNewComments component proper', () => {

    const component = rederComponent();

    expect(component).toMatchSnapshot();
  });

  test('TextBox should have the correct input', () => {
    const component = rederComponent();
    const textBox = component.getByTestId('comment_text');
    fireEvent.change(textBox, { target: { value: 'hello' } })
    expect((textBox as HTMLInputElement).value).toBe('hello')
  });

  test('save button click should have the correct input', () => {
    const component = rederComponent();

    React.useState = jest.fn().mockReturnValue(['hello', {}])

    const textBox = component.getByTestId('comment_text');

    fireEvent.change(textBox, { target: { value: 'hello' } });

    const button = component.getByTestId('save_comment');

    fireEvent.click(button)

    expect((textBox as HTMLInputElement).value).toBe('hello');
  });
});
