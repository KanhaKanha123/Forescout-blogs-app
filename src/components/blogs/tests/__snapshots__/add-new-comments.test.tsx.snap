// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewComments component renders AddNewComments component proper 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        backgroundcolor="#f4f4f4"
        class="sc-bdDrbm ieHSHu"
        height="200px"
        padding="15px"
        width="500px"
      >
        <input
          area-label="this is a add new comment textbox"
          class="sc-fuzrJS beTwst"
          data-testid="comment_text"
          placeholder="Add new comment"
          required=""
          type="text"
        />
        <div
          aria-label="Save new comment"
          backgroundcolor="#4792ef"
          class="sc-gtcAbF dPqWXl"
          data-testid="save_comment"
        >
          Save
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      backgroundcolor="#f4f4f4"
      class="sc-bdDrbm ieHSHu"
      height="200px"
      padding="15px"
      width="500px"
    >
      <input
        area-label="this is a add new comment textbox"
        class="sc-fuzrJS beTwst"
        data-testid="comment_text"
        placeholder="Add new comment"
        required=""
        type="text"
      />
      <div
        aria-label="Save new comment"
        backgroundcolor="#4792ef"
        class="sc-gtcAbF dPqWXl"
        data-testid="save_comment"
      >
        Save
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
